C++ API parity: Tanh

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27309 C++ API parity: MultiheadAttention
* #27382 C++ API parity: Linear
* #27538 C++ API parity: Threshold
* #27537 C++ API parity: Tanhshrink
* **#27536 C++ API parity: Tanh**

Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

pbelevich committed:
C++ API parity: Tanh

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"




[ghstack-poisoned]

yf225 commented on torch/csrc/api/include/torch/nn/functional/activation.h:
```diff
@@ -92,6 +92,11 @@ inline Tensor softsign(const Tensor& input) {
+inline Tensor tanh(const Tensor& input) {
+  TORCH_WARN("nn.functional.tanh is deprecated. Use torch.tanh instead.");
+  return input.tanh();
+}
```
We can probably just remove `F::tanh` in light of the deprecation message.

yf225 commented on torch/csrc/api/src/nn/modules/activation.cpp:
```diff
@@ -244,5 +244,17 @@ void SoftsignImpl::pretty_print(std::ostream& stream) const {
+// ============================================================================
+
+Tensor TanhImpl::forward(const Tensor& input) {
+  return input.tanh();
```
nit: the following would match the Python version even better:
```suggestion
  return torch::tanh(input);
```

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

yf225 reviewed (approved):
Thanks for the great work @pbelevich!

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Tanh"


Differential Revision: [D17835411](https://our.internmc.facebook.com/intern/diff/D17835411)

[ghstack-poisoned]
